File: src/dataset/python/saddle-points_exercism-ThomasZumsteg/claude/ClaudeSonnet4_saddle_points_v1.py
"""Finds saddle points in a rectangular 2d array"""

def saddle_points(matrix):
    """Finds points that are the largest in their row
    and smallest in their column
    """
    if not matrix:
        return set()
    
    # Validate matrix structure once
    row_length = len(matrix[0])
    if not all(len(row) == row_length for row in matrix):
        raise ValueError
    
    saddles = set()
    
    # Pre-compute all row maxima and column minima
    row_maxima = [max(row) for row in matrix]
    col_minima = [min(matrix[r][c] for r in range(len(matrix))) for c in range(row_length)]
    
    # Find saddle points by comparing pre-computed values
    for r in range(len(matrix)):
        row_max = row_maxima[r]
        for c in range(row_length):
            if row_max == col_minima[c]:
                saddles.add((r, c))
    
    return saddles

File: src/dataset/javascript/saddle-points_exercism-ThomasZumsteg/claude/ClaudeSonnet4_saddle-points_v1.js
var Matrix = function(matrix) {
	/* Two dimensional matrix class */
	this.rows = matrix.split("\n").map(function(row) { 
		return row.replace(/^\s+|\s+$/g,'').split(' ')
			.map(function(e) { return parseInt(e, 10)});
	});

	var numCols = this.rows[0].length;
	this.columns = new Array(numCols);
	for (var col = 0; col < numCols; col++) {
		this.columns[col] = new Array(this.rows.length);
		for (var row = 0; row < this.rows.length; row++) {
			this.columns[col][row] = this.rows[row][col];
		}
	}
	
	this.saddlePoints = saddlePoints(this);
}

function saddlePoints(matrix) {
	/* Finds saddle points in a matrix */
	var numRows = matrix.rows.length;
	var numCols = matrix.columns.length;
	
	var row_max = new Array(numRows);
	for (var r = 0; r < numRows; r++) {
		var max = matrix.rows[r][0];
		for (var c = 1; c < matrix.rows[r].length; c++) {
			if (matrix.rows[r][c] > max) {
				max = matrix.rows[r][c];
			}
		}
		row_max[r] = max;
	}
	
	var column_min = new Array(numCols);
	for (var c = 0; c < numCols; c++) {
		var min = matrix.columns[c][0];
		for (var r = 1; r < matrix.columns[c].length; r++) {
			if (matrix.columns[c][r] < min) {
				min = matrix.columns[c][r];
			}
		}
		column_min[c] = min;
	}
	
	var saddlePoints = [];
	for (var r = 0; r < numRows; r++) {
		for (var c = 0; c < matrix.rows[r].length; c++) {
			var element = matrix.rows[r][c];
			if (element === row_max[r] && element === column_min[c]) {
				saddlePoints.push([r, c]);
			}
		}
	}
	
	return saddlePoints;
};

module.exports = Matrix;